### PR TITLE
fix: various optimizations, idiomatic changes

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: various optimizations, idiomatic changes

## Motivation and Context
- acquire single leader lock for all background operations
  - healing, crawling and applying lifecycle policies.

- simplify lifecyle to avoid network calls, which was a
  bug in implementation - we should hold a leader and
  do everything from there, we have access to entire
  name space.

- make listing, walking not interfere by slowing itself
  down like the crawler.

- effectively use global context everywhere to ensure
  proper shutdown, in cache, lifecycle, healing

- don't read `format.json` for prometheus metrics in
  StorageInfo() call.

## How to test this PR?
Nothing changes, all functionality the same but many improvements.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [x] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#d7ed3bbd-6dcc-459d-8471-ffe8136300e3) |
<a name="d7ed3bbd-6dcc-459d-8471-ffe8136300e3"></a>
### 9179-4b3c934/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:30956
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4388e1bd47e2:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 7 minutes and 16 seconds
(4/15) Running aws-sdk-ruby tests ... done in 18 seconds
(5/15) Running awscli tests ... FAILED in 3 minutes and 19 seconds
{
  "name": "awscli",
  "duration": 1506,
  "function": "make_bucket_with_lock\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#8e85c1dc-ef08-4353-9558-d10e5f4fde06) |
<a name="8e85c1dc-ef08-4353-9558-d10e5f4fde06"></a>
### 9179-6906046/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:32176
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6d9c299b67da:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 9 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 7 minutes and 13 seconds
(4/15) Running aws-sdk-ruby tests ... done in 18 seconds
(5/15) Running awscli tests ... FAILED in 17 seconds
{
  "name": "awscli",
  "duration": 1515,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/posix.go [hunk 1]
```diff
@@ -672,11 +672,15 @@ func (s *posix) Walk(volume, dirPath, marker string, recursive bool, leafFile st
...
-	ch = make(chan FileInfo)
+	// buffer channel matches the S3 ListObjects implementation
+	ch = make(chan FileInfo, maxObjectList)
 	go func() {
 		defer close(ch)
 		listDir := func(volume, dirPath, dirEntry string) (emptyDir bool, entries []string) {
+			// Dynamic time delay.
+			t := UTCNow()
 			entries, err := s.ListDir(volume, dirPath, -1, leafFile)
+			sleepDuration(time.Since(t), 10.0)
```
Comment (vadmeste) [inline cmd/posix.go:683]: Not clear why this sleep, could you elaborate with a comment ?

File: cmd/daily-lifecycle-ops.go [hunk 2]
```diff
@@ -30,90 +30,37 @@ const (
...
-		objAPI = newObjectLayerWithoutSafeModeFn()
-		if objAPI == nil {
-			time.Sleep(time.Second)
-			continue
-		}
-		break
-	}
-
-	// Calculate the time of the last lifecycle operation in all peers node of the cluster
-	computeLastLifecycleActivity := func(status []BgOpsStatus) time.Time {
```
Comment (vadmeste) [inline cmd/daily-lifecycle-ops.go:68]: So this is a simplification of the code, if the leader node responsible of lifecycle background ops, the next node will do it immediately ?

**Review by vadmeste — COMMENTED** [review]
Some comments

File: cmd/posix.go [hunk 3]
```diff
@@ -672,11 +672,15 @@ func (s *posix) Walk(volume, dirPath, marker string, recursive bool, leafFile st
 		return nil, err
 	}
 
-	ch = make(chan FileInfo)
+	// buffer channel matches the S3 ListObjects implementation
+	ch = make(chan FileInfo, maxObjectList)
```
Comment (vadmeste) [inline cmd/posix.go:676]: By the way, a similar change in Walk() in storage-rest-client.go could also be efficient

File: cmd/posix.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/posix.go:683]: This is to wait if the listDir is taking longer it's like crawling, to slow down it calls are taking longer than tolerable limits.

File: cmd/daily-lifecycle-ops.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/daily-lifecycle-ops.go:68]: Next node will never do it, life cycle is dumb like crawling. Idea is that only one node is responsible..

File: cmd/posix.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/posix.go:676]: Yeah that's not done on purpose because there can be many disks to avoid memory build up.

File: cmd/posix.go (hunk 1 above)
Comment (vadmeste) [reply cmd/posix.go:683]: > This is to wait if the listDir is taking longer it's like crawling, to slow down it calls are taking longer than tolerable limits.

Are you going to merge this in master or in a separate test branch ?

Comment (harshavardhana) [reply cmd/posix.go:683]: This will be in master..

**Review by vadmeste — APPROVED** [review]
LGTM, tested healing/lifecycle

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**